### ISSUE TITLE
Evaluate effect of failed geo lookup on performance

### ISSUE THREAD

**Issue by aaronjudd** [reactioncommerce/reaction#1837]
Evaluate failure handling of geo lookups and ensure that if the geo server is down there isn't a lengthy timeout, and that errors are handling elegantly.  This code may be fine, but needs evaluation from a performance view.

**Comment by zenweasel** [comment]
There are two places where `shop/getLocale` is getting called, one on startup where it is called asynchronously and in the email sending code.

The call from email is a little confusing as the geoIP lookup is not really necessary, you would want to use shop settings I would think, but the geoIP lookup overrides that which would be the geolocation of the server not the user nor the store. So maybe we would want to wrap that code in a `isClient` block so we only get a geoIP lookup when we are called from the client and otherwise fallback to shop settings?

Just focusing on performance, this part is a blocking call (I believe these calls to `Reaction.Email` are before the job is queued). So it might be better to either just skip the geoIP lookup or at least put a timeout on it. Although the impact on performance would still probably be small.

I did find and remove the unused `getLocation` method which is what I think @aaronjudd and I were looking at the other day. This `getLocale` code is definitely used.

**Comment by zenweasel** [comment]
Closed via #2019

**Comment by zenweasel** [comment]
So apparently this is still causing a problem, because the GeoIP server is currently down and I'm not getting any prices on the site:


![reaction](https://user-images.githubusercontent.com/72819/29801878-8ccee0fe-8ca4-11e7-869c-10927c88937b.png)

**Comment by spencern** [comment]
@ticean any update on this issue?

**Comment by aaronjudd** [comment]
I just tested, reviewed code with @ticean and am unable to replicate the pricing behavior in 1.8.0.  While I do think that a anti-pattern has been introduced here, were the URL is hardcoded rather that using a setting or ENV variable, and we should create an issue to clean up that.  It doesn't appear that we're blocking anywhere either (getLocale is using this.unblock).  Not sure why we're doing direct calls via HTTP get rather than using a method either, but I don't see where the API would prevent address add, or pricing.  

Performance on the CLI is different issue, and probably not directly related to the Reaction handling.

**Comment by spencern** [comment]
Seems like we can close this ticket once we've created issues to add using an ENV var or setting to set the URL.

**Comment by ticean** [comment]
# GeoIP

### Setup

A GeoIP lookup client is configured in [server/api/geocoder.js](https://github.com/reactioncommerce/reaction/blob/f40ff536c139d70da02ca10ae12655247452d658/server/api/geocoder.js) and exported in [server/api/index.js](https://github.com/reactioncommerce/reaction/blob/db4e8d2aa500a9da59f994e64df4f8999ebe6965/server/api/index.js)

### Usage of GeoIP

1. [`server/methods/core/shop.js#L491-L521`](https://github.com/reactioncommerce/reaction/blob/f40ff536c139d70da02ca10ae12655247452d658/server/methods/core/shop.js#L491-L521)  
  GeoIP is used to determine users' location and autopopulate address fields in the address form.  
2. [`server/methods/core/shop.js#L204-L208`](https://github.com/reactioncommerce/reaction/blob/f40ff536c139d70da02ca10ae12655247452d658/server/methods/core/shop.js#L204-L208)  
  GeoIP is used to determine users' countryCode and locale.
3. [`imports/plugins/core/accounts/client/templates/addressBook/add/add.js#L13-L18`](https://github.com/reactioncommerce/reaction/blob/f40ff536c139d70da02ca10ae12655247452d658/imports/plugins/core/accounts/client/templates/addressBook/add/add.js#L13-L18)  
  Remote GeoIP is called ***manually*** on address book additions and is used to set `country_code`.  
4. [`server/methods/accounts/accounts.js#L265-L301`](https://github.com/reactioncommerce/reaction/blob/f40ff536c139d70da02ca10ae12655247452d658/server/methods/accounts/accounts.js#L265-L301)  
  Used to validate an address. Can be overridden with a custom validator.  
5. [`reaction/reaction-cli#/src/utils/geo.js`](https://github.com/reactioncommerce/reaction-cli/blob/d88d673fbf3e48443a054950c62e55dcf4f61938/src/utils/geo.js) 
  Reaction CLI uses a GeoIP function that's used in [analytics](https://github.com/reactioncommerce/reaction-cli/blob/37df6178702fe4071464f46050b25be8045597de/src/utils/analytics.js) when a 
  Segment key is not present.
  
### Affects of Failed Geo Lookup

In most cases this will not have an impact. The calls are async and will eventually time out without blocking.

In some cases, if the server is latent then the response could happen after user input. This would be the worst case.

---

As @aaronjudd mentioned, there are some hardcoded references to the geo server. I'll create a new issue for that. In those cases the default store locale settings will be used.

**Comment by spencern** [comment]
Closing in favor of #3655
